storage/remote/metadata_watcher.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/scrape"
+)
+
+// MetadataAppender is an interface used by the Metadata Watcher to send metadata read from the scrape manager on to somewhere else.
+type MetadataAppender interface {
+	AppendMetadata(context.Context, []scrape.MetricMetadata)
+}
+
+// Watchable is an interface used to represent a scrape manager from which we fetch metadata.
+type Watchable interface {
+	TargetsActive() map[string][]*scrape.Target
+}
+
+// MetadataWatcher watches the Scrape Manager for a given WriteMetadataTo.
+type MetadataWatcher struct {
+	name   string
+	logger log.Logger
+
+	managerGetter scrape.ManagerGetter
+	manager       Watchable
+	writer        MetadataAppender
+
+	interval model.Duration
+	deadline time.Duration
+
+	quit chan struct{}
+	done chan struct{}
+
+	hardShutdown context.CancelFunc
+}
+
+// NewMetadataWatcher builds a new MetadataWatcher.
+func NewMetadataWatcher(l log.Logger, mg scrape.ManagerGetter, name string, w MetadataAppender, interval model.Duration, deadline time.Duration) *MetadataWatcher {
+	if l == nil {
+		l = log.NewNopLogger()
+	}
+
+	if mg == nil {
+		mg = &scrape.ReadyManager{}
+	}
+
+	return &MetadataWatcher{
+		name:   name,
+		logger: l,
+
+		managerGetter: mg,
+		writer:        w,
+
+		interval: interval,
+		deadline: deadline,
+
+		quit: make(chan struct{}),
+		done: make(chan struct{}),
+	}
+}
+
+// Start the MetadataWatcher.
+func (mw *MetadataWatcher) Start() {
+	level.Info(mw.logger).Log("msg", "Starting scraped metadata watcher")
+	var hardShutdownCtx context.Context
+	hardShutdownCtx, mw.hardShutdown = context.WithCancel(context.Background())
+
+	go mw.loop(hardShutdownCtx)
```

Comment:
My preference would be to either stick to closing channels or context, having both might be confusing for a reader. Also again, since loop is part of `MetadataWatcher` it looks odd that we pass to loop `hardShutdownCtx` but not `mw.quit` and `mw.done`. Can we stick to one way of doing the thing? 

I would say:
* m.hardShutdownCtx
* m.hardShutdownCancel
* m.softShutdownCtx
* m.softShutfownCancel
* m.done chan struct (alternatively sync.WaitGroup)

Then if you use all from struct (instead of passing some context by arg) it might be more clear? :thinking:

Reply:
I love this! Thank you very much for the suggestion 😍 🚀